scripts/ta_bin_to_c.py

```diff
@@ -22,6 +22,9 @@ def get_args():
     parser.add_argument('--out', required=True,
                         help='Name of the output C file')
 
+    parser.add_argument('--enumerate', type=int, required=True,
```

Comment:
There are TAs  which should be enumerated and TAa which should not. Both of them use the same python script to generate C code.

Reply:
How can one tell if a TA should be enumerated or not when calling this script?

Reply:
I think what @jenswi-linaro means in his first comment is, just make sure the (early) TA has the `TA_FLAG_DEVICE_ENUM` flag in its source code. Jens, correct me if I'm wrong ;-)
I would prefer that too.

Reply:
From the patch use EARLY_TA_PATHS=file.ta  if this TA should not be enumerated (visible to kernel) and use  EARLY_TA_ENUM_PATHS=file.ta is you want kernel to see this ta:
```
$(foreach f, $(EARLY_TA_PATHS), $(eval $(call process_early_ta,$(f),0)))
$(foreach f, $(EARLY_TA_ENUM_PATHS), $(eval $(call process_early_ta,$(f),1)))
scripts/ta_bin_to_c.py  --enumerate $(2)
```
@jenswi-linaro, @jforissier , this can not be flag inside TA source code. Optee-os need to register this TA with some flags before use it. So or do something like add additional elf section with flags or use variable to enum. As we discussed with Sumit  EARLY_TA_ENUM_PATHS looks like more clean solution.

Reply:
The script could extract the flag from `ta_head.flags`. You'd need to pars the ELF a bit, but we're already depending on `elftools` in other scripts.

Reply:
@jenswi-linaro  yes, that has to work also. I specially put flag changes to separate patch in case if it needs to be reconsidered. My proposal has 2 props 1) it can be easily extended to support other flags. 2) ta binary is the same is you want to built-in it or put on file system 3) builds controls which ta it wants to compile in and how. 4) change flags enum in optee os does not effect on ta souce code. 
I don't know the real deployment chain for this project but let's consider situation when customer provides TA only in binary form for your os distribution. You optee-os version can be changed and enum for flags compiled into TA might not match. Then you might want to see this TA on optee bus or might not. Or might be some other integration flags can be added in future. All solutions has props and cons I don't mind to change this patch, but please confirm that it is needed.

Reply:
Hmmmm I'm not convinced. Let's examine your arguments:

> 1) it can be easily extended to support other flags

Other flags can be added to `struct ta_head::ta_flags` too, I don't really see the point

> 2) ta binary is the same is you want to built-in it or put on file system

Same if we choose to add `TA_FLAG_DEVICE_ENUM` to `ta_head::flags`. It should not prevent the TA from running normally.

> 3) builds controls which ta it wants to compile in and how.

OK, why wouldn't it be the case with any other solution? ;-)

> 4) change flags enum in optee os does not effect on ta souce code.

True, but I'm not sure it is a significant issue. Note, it is always possible to change the flags in a compiled TA ELF. So if we want to be able to add an existing TA as an "enumerable early TA", we could easily do that in `ta_bin_to_c.py` too.

On the other hand, the drawbacks I see with your solution are:

a) It introduces a compile-time setting (`EARLY_TA_ENUM_PATHS`) which could be avoided. In addition, the content of `EARLY_TA_PATHS` does not represent all the early TAs anymore (this last part could be fixed however)
b) `TA_FLAG_DEVICE_ENUM` is now used in a different place than `struct ta_head::flags` (namely, `struct early_ta::flags`) and has no effect in an early TA's `struct ta_head::flags`. This is misleading and error-prone both from a usage (TA developper) and maintenance perspective.

FYI, what Jens mentioned (read the TA flags from the ELF) is really easy to do. For example:
```python
#!/usr/bin/env python3

from elftools.elf.elffile import ELFFile
import struct
import sys


def get_ta_flags(path):
    elffile = ELFFile(path)

    for s in elffile.iter_sections():
        if s.name == '.ta_head':
            return struct.unpack('<16x4xI', s.data()[:24])[0]

    raise Exception('.ta_head section not found')
            

def main():
    with open(sys.argv[1], 'rb') as f:
        print('0x{:04x}'.format(get_ta_flags(f)))


if __name__ == "__main__":
    main()
```
```
$ ./scripts/get_ta_flags.py out/arm/ta/pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee.stripped.elf 
0x001c
```

Reply:
Ok I will send update. Is there any  Dockerfile with dependencies? I have some issues to build the latest code under our OE build. Which python crypto module is needed python3-pycryptodome-native?

Reply:
> Ok I will send update. Is there any Dockerfile with dependencies? I have some issues to build the latest code under our OE build. 

Our [Shippable CI script](https://github.com/OP-TEE/optee_os/blob/3.8.0/.shippable.yml#L7) uses this container: https://hub.docker.com/repository/docker/jforissier/optee_os_ci_clangbuilt.
The same without the `_clangbuilt` suffix is, well, the same without a locally built Clang :wink: and should work equally well if you don't use Clang.

> Which python crypto module is needed python3-pycryptodome-native?

Honestly I don't know, this has been discussed so many times, in fact the package names are inconsistent from one distribution to another and even from one version to another... :cry:

Reply:
@jenswi-linaro @jforissier 
> The script could extract the flag from ta_head.flags. You'd need to pars the ELF a bit, but we're already depending on elftools in other scripts.
> FYI, what Jens mentioned (read the TA flags from the ELF) is really easy to do. For example:

I agree with your suggestions to use elf parsing in scripts (TBH, I wasn't aware of this feature).

My initial thoughts were to parse early TA `struct ta_head::flags` at runtime but that would require loading every early TA inside a loop to parse `flags` from runtime TA context. But that felt like much of overhead which can be avoided at build time instead via passing the flags via `struct early_ta::flags`. So adding `EARLY_TA_ENUM_PATHS` macro was an option to achieve this.

But looking at inconsistencies that @jforissier mentioned, I agree it would be better to keep a single copy of flags here: `struct ta_head::flags` and just sync them with `struct early_ta::flags` in case of early TAs.

Since we are on this topic, I would also like to discuss about enumeration of REE-FS TAs via tee-supplicant too. I guess we can have two options there as follows:

1. Either to sync `struct ta_head::flags` to new field: `struct shdr::flags` and tee-supplicant could parse that and report TA UUIDs which can be enumerated. But this might make the TA header backwards incompatible.
2. Or just put the TAs which could be enumerated under `/lib/optee-armtz/enum` folder. But that again would be inconsistent as per @jforissier comments.

Reply:
The solution must be backwards compatible.
We could split the `struct shdr::img_type` field into two 16-bit fields instead in such a way that it's backwards compatible. One of the fields remains as `img_type` and the is turned into a flags field. We should probably define special flags for this field instead of mirroring `TA_FLAG_*` in order to conserve bit usage.

Old TAs would have all bits zero in the flags field.

Reply:
@b49020 
> I agree it would be better to keep a single copy of flags here: `struct ta_head::flags` and just sync them with `struct early_ta::flags` in case of early TAs.

Exactly, copy the whole flags not just the enumerate bit.

> Since we are on this topic, I would also like to discuss about enumeration of REE-FS TAs via tee-supplicant too. I guess we can have two options there as follows:
> 
> 1. Either to sync `struct ta_head::flags` to new field: `struct shdr::flags` and tee-supplicant could parse that and report TA UUIDs which can be enumerated. But this might make the TA header backwards incompatible.
> 2. Or just put the TAs which could be enumerated under `/lib/optee-armtz/enum` folder. But that again would be inconsistent as per @jforissier comments.

I understand it's a bit annoying to rely on flags that are deep inside the TA, but here again I would advocate solution 1. As for compatibility, there must be a way, we have a magic number in `struct shdr`...

Reply:
IIUC, via backwards compatibility we mean to be able to run TAs which were built with older versions of `optee_os` only. But we don't support a combination like a machine running an older version of `optee_os` to run a TA built with recent version of `optee_os`, correct?

If that's the case then I think solution suggested by @jenswi-linaro sounds more reasonable to cater to future extensions as well.

Reply:
Right, we should be able to run older TAs.